templates/aws-vpc.template.yaml

```diff
@@ -303,849 +345,911 @@ Parameters:
+          - !Equals [!Ref CreatePrivateSubnets, "true"]
+          - !Equals [!Ref CreatePublicSubnets, "true"]
+        AssertDescription: At least one of CreatePublicSubnets or CreatePrivateSubnets must be set to 'true'
 Conditions:
   3AZCondition: !Or
     - !Equals
-      - !Ref 'NumberOfAZs'
-      - '3'
-    - !Condition '4AZCondition'
+      - !Ref "NumberOfAZs"
+      - "3"
+    - !Condition "4AZCondition"
   4AZCondition: !Equals
-    - !Ref 'NumberOfAZs'
-    - '4'
+    - !Ref "NumberOfAZs"
+    - "4"
   AdditionalPrivateSubnetsCondition: !And
     - !Equals
-      - !Ref 'CreatePrivateSubnets'
-      - 'true'
+      - !Ref "CreatePrivateSubnets"
+      - "true"
     - !Equals
-      - !Ref 'CreateAdditionalPrivateSubnets'
-      - 'true'
+      - !Ref "CreateAdditionalPrivateSubnets"
+      - "true"
   AdditionalPrivateSubnets&3AZCondition: !And
-    - !Condition 'AdditionalPrivateSubnetsCondition'
-    - !Condition '3AZCondition'
+    - !Condition "AdditionalPrivateSubnetsCondition"
+    - !Condition "3AZCondition"
   AdditionalPrivateSubnets&4AZCondition: !And
-    - !Condition 'AdditionalPrivateSubnetsCondition'
-    - !Condition '4AZCondition'
+    - !Condition "AdditionalPrivateSubnetsCondition"
+    - !Condition "4AZCondition"
+
+  AdditionalPrivateSubnets&PublicSubnets&NatGatewaysCondition: !And
+    - !Condition "AdditionalPrivateSubnetsCondition"
+    - !Condition "PublicSubnetsCondition"
+    - !Condition "NATGatewaysCondition"
+  AdditionalPrivateSubnets&PublicSubnets&NatGateways&3AZCondition: !And
+    - !Condition "AdditionalPrivateSubnets&3AZCondition"
+    - !Condition "PublicSubnetsCondition"
+    - !Condition "NATGatewaysCondition"
+  AdditionalPrivateSubnets&PublicSubnets&NatGateways&4AZCondition: !And
+    - !Condition "AdditionalPrivateSubnets&4AZCondition"
+    - !Condition "PublicSubnetsCondition"
+    - !Condition "NATGatewaysCondition"
+
   GovCloudCondition: !Equals
-    - !Ref 'AWS::Region'
+    - !Ref "AWS::Region"
     - us-gov-west-1
+  NATGatewaysCondition: !Equals
+    - !Ref "CreateNATGateways"
+    - "true"
+  NATGateways&3AZCondition: !And
+    - !Condition "NATGatewaysCondition"
+    - !Condition "3AZCondition"
+  NATGateways&4AZCondition: !And
+    - !Condition "NATGatewaysCondition"
+    - !Condition "4AZCondition"
   NVirginiaRegionCondition: !Equals
-    - !Ref 'AWS::Region'
+    - !Ref "AWS::Region"
     - us-east-1
   PrivateSubnetsCondition: !Equals
-    - !Ref 'CreatePrivateSubnets'
-    - 'true'
+    - !Ref "CreatePrivateSubnets"
+    - "true"
   PrivateSubnets&3AZCondition: !And
-    - !Condition 'PrivateSubnetsCondition'
-    - !Condition '3AZCondition'
+    - !Condition "PrivateSubnetsCondition"
+    - !Condition "3AZCondition"
   PrivateSubnets&4AZCondition: !And
-    - !Condition 'PrivateSubnetsCondition'
-    - !Condition '4AZCondition'
+    - !Condition "PrivateSubnetsCondition"
+    - !Condition "4AZCondition"
+  PrivateSubnets&PublicSubnets&NatGatewaysCondition: !And
```

Comment:
Whereever you are using PrivateSubnet&Publicsubnet&Natgateway, you can just use Natgateway condition, because assertion will make sure that public and private subnet are true when natgateway is true.